library/std/src/Std/Intrinsic.qs

```diff
@@ -1140,6 +1140,58 @@ operation Z(qubit : Qubit) : Unit is Adj + Ctl {
     adjoint self;
 }
 
+/// # Summary
+/// Applies the given unitary matrix to the given qubits. The matrix must be square, with dimensions dictated by the number of qubits.
+/// This operation is simulator-only and is not supported on hardware.
+///
+/// # Input
+/// ## matrix
+/// The unitary matrix to apply.
+/// ## qubits
+/// The qubits to which the unitary matrix should be applied.
+///
+/// # Remarks
+/// The matrix is checked at runtime to ensure it's shape is square and that the matrix dimensions are `2 ^ Length(qubits)`.
+///
+/// # Example
+/// This performs a two qubit CNOT using the unitary matrix representation:
+/// ```qsharp
+/// import Std.Math.Complex;
+/// use qs = Qubit[2];
+/// let one = new Complex { Real = 1.0, Imag = 0.0 };
+/// let zero = new Complex { Real = 0.0, Imag = 0.0 };
+/// ApplyUnitary(
+///     [
+///         [one, zero, zero, zero],
+///         [zero, one, zero, zero],
+///         [zero, zero, zero, one],
+///         [zero, zero, one, zero]
+///     ],
+///     qs
+/// );
+/// ```
+@Config(Unrestricted)
+operation ApplyUnitary(matrix : Complex[][], qubits : Qubit[]) : Unit {
+    let num_rows = Length(matrix);
+    for col in matrix {
+        if Length(col) != num_rows {
+            fail "matrix passed to ApplyUnitary must be square.";
+        }
+    }
+
+    let num_qubits = Length(qubits);
+    if num_rows != 1 <<< num_qubits {
+        fail "matrix passed to ApplyUnitary must have dimensions 2^Length(qubits).";
+    }
+
+    Apply(matrix, qubits);
+}
+
+@Config(Unrestricted)
+operation Apply(matrix : Complex[][], qubits : Qubit[]) : Unit {
```

Comment:
I like "Apply" as a name, but on the other hand "Apply" is a very generic name. I'm afraid it could be easily confused or conflict with something else... Could we make it more specific? If not, it's fine.

Reply:
This one is not exposed to the user, only `ApplyUnitary` is exported. So this is effectively an internal utility. At first I had it called `ApplyUnitaryInternal` but then noticed that circuits will display this name, since it's the intrinsic, and not the wrapper, and it looked weird to have `ApplyUnitaryInternal` in the circuit diagram. So I settled at something shorter that looks better for circuits. The other bit of reasoning that I had was that every intrinsic needs to globally unique in its name so it seemed better to pick something non-specific as that felt less likely to overlap with a custom intrinsic someone might try to use later.